### PR TITLE
CO-1120: sanity check sources

### PR DESCRIPTION
**Pull request by ianbrown9475**

### CODE REVIEW

File: src/main/groovy/edu/oregonstate/mist/locations/MergeUtil.groovy [hunk 1]
```diff
@@ -162,5 +163,17 @@ class MergeUtil {
             }
         }
 
+        def missingLocations = []
+        campusMapData.each { key, value ->
+            if(!data.find { it.id == value.id }) {
+                missingLocations.add(value.id)
+            }
+        }
+        float ratioMissing = missingLocations.size() / campusMapData.size()
```
Comment (jaredkosanovic) [inline src/main/groovy/edu/oregonstate/mist/locations/MergeUtil.groovy]: Could you log `ratioMissing` to `info` every time (even if it doesn't exceed the threshold)?

File: src/main/groovy/edu/oregonstate/mist/locations/MergeUtil.groovy [hunk 2]
```diff
@@ -162,5 +163,17 @@ class MergeUtil {
...
+        if(ratioMissing < MISSING_LOCATIONS_THRESHOLD) {
```
Comment (jaredkosanovic) [inline src/main/groovy/edu/oregonstate/mist/locations/MergeUtil.groovy]: Could you add a space between if and ( in your if statements to match the style of the other if statements in this file?

File: src/main/groovy/edu/oregonstate/mist/locations/MergeUtil.groovy [hunk 3]
```diff
@@ -162,5 +163,17 @@ class MergeUtil {
             }
         }
 
+        def missingLocations = []
+        campusMapData.each { key, value ->
```
Comment (jaredkosanovic) [inline src/main/groovy/edu/oregonstate/mist/locations/MergeUtil.groovy]: I think you might be able to shorten this to something like this.
`def missingLocations = campusMapData.keySet() - data.collect { it.id }`

File: src/main/groovy/edu/oregonstate/mist/locations/db/ArcGisDAO.groovy [hunk 4]
```diff
@@ -48,8 +50,11 @@ class ArcGisDAO {
         def mappedData = mapper.readTree(gisData).get("features")
         def data = new HashMap<String, GenderInclusiveRRLocation>()
 
-        if (mappedData.size() == 0) {
-            throw new DAOException("found zero gender inclusive restrooms")
+        int numFound = mappedData.size()
+
+        if (numFound < ARCGIS_THRESHOLD) {
```
Comment (jaredkosanovic) [inline src/main/groovy/edu/oregonstate/mist/locations/db/ArcGisDAO.groovy]: You have a lot of these checks which are pretty similar. They compare the size of two integers, and throw an exception if one is less than the other. You could probably put this check in a static shared method somewhere?

**Review by jaredkosanovic — CHANGES_REQUESTED** [review]
Nice work so far

File: src/main/groovy/edu/oregonstate/mist/locations/MergeUtil.groovy (hunk 1 above)
Comment (ianbrown9475) [reply src/main/groovy/edu/oregonstate/mist/locations/MergeUtil.groovy]: Definitely a good idea. I think this might be helpful in determining what the threshold should be in the first place.

**Comment by ianbrown9475** [comment]
Realized my logic was mixed in places so I just want to clarify the missingLocationsRatio field. This should be the maximum ratio of locations in campus maps that are not found in our sources. It should be something like 0.05, 0.1, or 0.15, similar to a margin of error.

**Comment by jaredkosanovic** [comment]
> This should be the maximum ratio of locations in campus maps that are not found in our sources. It should be something like 0.05, 0.1, or 0.15, similar to a margin of error.

Good catch. Could you add something similar to your comment in the configuration example?

Also, what do you think about adding some unit tests for some of these sanity checks?

**Review by jaredkosanovic — APPROVED** [review]
